File: apps/api/src/routers/characters.py
"""Character selection API endpoints.

Provides endpoints for:
- GET /api/characters - List all available characters (public, no auth)
- GET /api/users/character - Get current user's selected character (auth required)
- PUT /api/users/character - Update user's selected character (auth required)
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.auth.middleware import get_current_user
from src.auth.schemas import FirebaseUser
from src.database import get_db
from src.repositories.email_repository import get_user_by_firebase_uid
from src.schemas.character import (
    CharacterResponse,
    CharactersListResponse,
    UpdateCharacterRequest,
)
from src.services.character_service import get_all_characters, get_character

router = APIRouter()


@router.get("/characters", response_model=CharactersListResponse)
async def list_characters() -> CharactersListResponse:
    """Return all available characters.

    This endpoint is public and does not require authentication.
    """
    characters = get_all_characters()
    return CharactersListResponse(
        characters=[
            CharacterResponse(
                id=c.id,
                displayName=c.display_name,
                description=c.description,
            )
            for c in characters
        ]
    )


@router.get("/users/character", response_model=CharacterResponse)
async def get_user_character(
    current_user: FirebaseUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> CharacterResponse:
    """Return the current user's selected character.

    Falls back to default character (gyaru) if no selection.
    """
    user = await get_user_by_firebase_uid(session, current_user.uid)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail={"error": "user_not_found"},
        )

    character = get_character(user.selected_character_id)
    return CharacterResponse(
        id=character.id,
        displayName=character.display_name,
        description=character.description,
    )


@router.put("/users/character", response_model=CharacterResponse)
async def update_user_character(
    request: UpdateCharacterRequest,
    current_user: FirebaseUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> CharacterResponse:
    """Update the current user's selected character.

    Returns 400 if the character ID is invalid.
    """
    # Validate character ID
    all_ids = [c.id for c in get_all_characters()]
    if request.characterId not in all_ids:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"error": "invalid_character_id"},
        )

    user = await get_user_by_firebase_uid(session, current_user.uid)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail={"error": "user_not_found"},
        )

    user.selected_character_id = request.characterId
    await session.commit()

    character = get_character(request.characterId)
    return CharacterResponse(
        id=character.id,
        displayName=character.display_name,
        description=character.description,
    )

File: apps/web/src/api/characters.ts
/**
 * Characters API functions.
 */

const API_BASE_URL = '/api';

/**
 * Character data from API.
 */
export interface Character {
  id: string;
  displayName: string;
  description: string;
}

/**
 * API response for GET /api/characters.
 */
export interface CharactersListResponse {
  characters: Character[];
}

/**
 * Fetch all available characters (no auth required).
 *
 * @returns Promise resolving to CharactersListResponse
 * @throws Error if the request fails
 */
export async function fetchCharacters(): Promise<CharactersListResponse> {
  const response = await fetch(`${API_BASE_URL}/characters`, {
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail?.error || 'キャラクター一覧の取得に失敗しました');
  }

  return response.json();
}

/**
 * Fetch current user's selected character.
 *
 * @param idToken - Firebase ID token for authentication
 * @returns Promise resolving to Character
 * @throws Error if the request fails
 */
export async function fetchCurrentCharacter(idToken: string): Promise<Character> {
  const response = await fetch(`${API_BASE_URL}/users/character`, {
    headers: {
      Authorization: `Bearer ${idToken}`,
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail?.error || '現在のキャラクターの取得に失敗しました');
  }

  return response.json();
}

/**
 * Update user's selected character.
 *
 * @param idToken - Firebase ID token for authentication
 * @param characterId - ID of the character to select
 * @returns Promise resolving to updated Character
 * @throws Error if the request fails
 */
export async function updateCharacter(idToken: string, characterId: string): Promise<Character> {
  const response = await fetch(`${API_BASE_URL}/users/character`, {
    method: 'PUT',
    headers: {
      Authorization: `Bearer ${idToken}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ characterId }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail?.error || 'キャラクターの更新に失敗しました');
  }

  return response.json();
}
